File: sdk/flutter/android/src/main/kotlin/com/openstory/sdk/flutter/OpenStoryFlutterCallbacks.kt
package com.openstory.sdk.flutter

import com.openstory.sdk.OpenStoryCallbacks
import com.openstory.sdk.model.OpenStoryAnalyticsEvent
import com.openstory.sdk.model.OpenStoryCtaPayload

internal class OpenStoryFlutterCallbacks(
    private val eventStreamHandler: BufferedEventStreamHandler,
) : OpenStoryCallbacks {
    override fun onStoryBarImpression(event: OpenStoryAnalyticsEvent) {
        eventStreamHandler.send(event.toFlutterPayload())
    }

    override fun onStoryGroupTap(event: OpenStoryAnalyticsEvent) {
        eventStreamHandler.send(event.toFlutterPayload())
    }

    override fun onStoryView(event: OpenStoryAnalyticsEvent) {
        eventStreamHandler.send(event.toFlutterPayload())
    }

    override fun onStoryComplete(event: OpenStoryAnalyticsEvent) {
        eventStreamHandler.send(event.toFlutterPayload())
    }

    override fun onStoryCtaTap(payload: OpenStoryCtaPayload) {
        eventStreamHandler.send(payload.toFlutterPayload())
    }

    override fun onViewerClose(event: OpenStoryAnalyticsEvent) {
        eventStreamHandler.send(event.toFlutterPayload())
    }

    override fun onGroupComplete(event: OpenStoryAnalyticsEvent) {
        eventStreamHandler.send(event.toFlutterPayload())
    }

    override fun onError(
        placementKey: String,
        throwable: Throwable,
    ) {
        eventStreamHandler.send(
            mapOf(
                "type" to "error",
                "placementKey" to placementKey,
                "message" to (
                    throwable.message
                        ?: throwable::class.java.simpleName
                        ?: "Unknown OpenStory error."
                    ),
                "errorType" to throwable::class.java.name,
            ),
        )
    }
}

private fun OpenStoryAnalyticsEvent.toFlutterPayload(): Map<String, Any?> {
    return mapOf(
        "type" to "analytics",
        "kind" to kind.toWireValue(),
        "placementKey" to placementKey,
        "storyGroupId" to storyGroupId,
        "storyGroupRevisionId" to storyGroupRevisionId,
        "storyId" to storyId,
        "storyRevisionId" to storyRevisionId,
        "occurredAtMillis" to occurredAtMillis,
    )
}

private fun OpenStoryAnalyticsEvent.Kind.toWireValue(): String {
    return when (this) {
        OpenStoryAnalyticsEvent.Kind.STORY_BAR_IMPRESSION -> "story_bar_impression"
        OpenStoryAnalyticsEvent.Kind.STORY_GROUP_TAP -> "story_group_tap"
        OpenStoryAnalyticsEvent.Kind.STORY_VIEW -> "story_view"
        OpenStoryAnalyticsEvent.Kind.STORY_COMPLETE -> "story_complete"
        OpenStoryAnalyticsEvent.Kind.STORY_CTA_TAP -> "story_cta_tap"
        OpenStoryAnalyticsEvent.Kind.VIEWER_CLOSE -> "viewer_close"
        OpenStoryAnalyticsEvent.Kind.GROUP_COMPLETE -> "group_complete"
    }
}

private fun OpenStoryCtaPayload.toFlutterPayload(): Map<String, Any?> {
    return mapOf(
        "type" to "cta",
        "placementKey" to placementKey,
        "storyGroupId" to storyGroupId,
        "storyGroupRevisionId" to storyGroupRevisionId,
        "storyId" to storyId,
        "storyRevisionId" to storyRevisionId,
        "label" to label,
        "targetType" to targetType.toWireValue(),
        "targetValue" to targetValue,
    )
}

private fun OpenStoryCtaPayload.TargetType.toWireValue(): String {
    return when (this) {
        OpenStoryCtaPayload.TargetType.URL -> "url"
        OpenStoryCtaPayload.TargetType.DEEPLINK -> "deeplink"
    }
}

File: sdk/flutter/ios/Classes/OpenStoryFlutterCallbacks.swift
import Flutter
import Foundation

final class OpenStoryEventStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var bufferedEvents = [[String: Any]]()
    private var closed = false

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        let pendingEvents = bufferedEvents
        bufferedEvents.removeAll()
        pendingEvents.forEach { events($0) }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    func send(_ event: [String: Any]) {
        guard !closed else {
            return
        }

        if let eventSink {
            eventSink(event)
            return
        }

        bufferedEvents.append(event)
    }

    func close() {
        closed = true
        eventSink = nil
        bufferedEvents.removeAll()
    }
}

@MainActor
final class OpenStoryFlutterCallbacks: NSObject, OpenStoryCallbacks {
    private let streamHandler: OpenStoryEventStreamHandler

    init(streamHandler: OpenStoryEventStreamHandler) {
        self.streamHandler = streamHandler
        super.init()
    }

    func onStoryBarImpression(event: OpenStoryAnalyticsEvent) {
        streamHandler.send(analyticsPayload(from: event))
    }

    func onStoryGroupTap(event: OpenStoryAnalyticsEvent) {
        streamHandler.send(analyticsPayload(from: event))
    }

    func onStoryView(event: OpenStoryAnalyticsEvent) {
        streamHandler.send(analyticsPayload(from: event))
    }

    func onStoryComplete(event: OpenStoryAnalyticsEvent) {
        streamHandler.send(analyticsPayload(from: event))
    }

    func onStoryCtaTap(payload: OpenStoryCtaPayload) {
        streamHandler.send(ctaPayload(from: payload))
    }

    func onViewerClose(event: OpenStoryAnalyticsEvent) {
        streamHandler.send(analyticsPayload(from: event))
    }

    func onGroupComplete(event: OpenStoryAnalyticsEvent) {
        streamHandler.send(analyticsPayload(from: event))
    }

    func onError(placementKey: String, error: any Error) {
        let nsError = error as NSError
        var payload: [String: Any] = [
            "type": "error",
            "placementKey": placementKey,
            "message": nsError.localizedDescription,
            "code": String(nsError.code),
            "errorType": String(describing: type(of: error)),
        ]
        if payload["message"] as? String == nil || (payload["message"] as? String)?.isEmpty == true {
            payload["message"] = "Unknown OpenStory error."
        }
        streamHandler.send(payload)
    }

    private func analyticsPayload(from event: OpenStoryAnalyticsEvent) -> [String: Any] {
        var payload: [String: Any] = [
            "type": "analytics",
            "kind": event.kind.rawValue,
            "placementKey": event.placementKey,
            "occurredAtMillis": event.occurredAtMillis,
        ]

        if let storyGroupId = event.storyGroupId {
            payload["storyGroupId"] = storyGroupId
        }
        if let storyGroupRevisionId = event.storyGroupRevisionId {
            payload["storyGroupRevisionId"] = storyGroupRevisionId
        }
        if let storyId = event.storyId {
            payload["storyId"] = storyId
        }
        if let storyRevisionId = event.storyRevisionId {
            payload["storyRevisionId"] = storyRevisionId
        }

        return payload
    }

    private func ctaPayload(from payload: OpenStoryCtaPayload) -> [String: Any] {
        [
            "type": "cta",
            "placementKey": payload.placementKey,
            "storyGroupId": payload.storyGroupId,
            "storyGroupRevisionId": payload.storyGroupRevisionId,
            "storyId": payload.storyId,
            "storyRevisionId": payload.storyRevisionId,
            "label": payload.label,
            "targetType": payload.targetType.rawValue,
            "targetValue": payload.targetValue,
        ]
    }
}
